### PR TITLE
Adding task that runs eslint as configured in the app checkup is being run on

### PR DESCRIPTION
**Pull request by carakessler**
This run eslint as configured in the app checkup is being run on, and summarizes the results.

### CODE REVIEW

File: packages/checkup-plugin-javascript/__tests__/eslint-summary-task-test.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    project.writeSync();
+
+    const eslintConfig = readEslintConfig(project.filePaths, project.baseDir, pkg);
+
+    expect(eslintConfig).toEqual(expectedConfig);
+    project.dispose();
+  });
+});
+
+function filterResultFilePath(report: CLIEngine.LintReport) {
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/__tests__/eslint-summary-task-test.ts:130]: We have a specific type we export from core for this so we can keep the type defs centralized.

```ts
export type ESLintReport = CLIEngine.LintReport;
```

File: packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  failureCount: number;
+  fixableFailureCount: number;
+};
+
+type EslintDataForConsole = {
+  errorCount: number;
+  warningCount: number;
+  errorList: { rule: string; failures: string }[];
+  warningList: { rule: string; failures: string }[];
+};
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:15]: Why is this type only for console?

File: packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+export default class EslintSummaryTaskResult extends BaseTaskResult implements TaskResult {
+  esLintReport!: ESLintReport;
+
+  toConsole() {
+    let sortedEslintReport = transformEslintReport(this.esLintReport);
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:21]: I'd love it if we could consolidate these transform functions around a pattern that ppl could follow vs. ad hoc functions we make. Even having a function interface would at least allow authors to follow a well trodden path. Nothing necessarily we should do right now, but we should definitely be thinking about it.

File: packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  let errors: Record<string, EslintFailureCounts> = {};
+  let warnings: Record<string, EslintFailureCounts> = {};
+
+  // we want to strip these from total counts (generally they are just a warning that a file was ignored)
+  let warningsWithoutRulesIds = 0;
+
+  esLintReport.results.forEach((result) => {
+    if (result.messages.length > 0) {
+      result.messages.forEach((msg) => {
+        // severity 1 is a warning
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:62]: Should you create an enum so this reads better without the need for a comment?

File: packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+function createFailureList(record: Record<string, EslintFailureCounts>, failureType: string) {
+  let chalkColor: Function = failureType === 'errors' ? chalk.red : chalk.yellow;
+
+  return Object.entries(record).map((arr) => {
+    return {
+      rule: arr[0],
+      failures: `${chalkColor(arr[1].failureCount + arr[1].fixableFailureCount)} ${failureType} (${
+        arr[1].fixableFailureCount
+      } fixable)`,
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:110]: Can you store the value of these array indexes so we can make more sense of what the values are?

File: packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+ * @description In prioritized order as specified by https://eslint.org/docs/user-guide/configuring
+ */
+export const ACCEPTED_ESLINT_CONFIG_FILES = [
+  '.eslintrc.js',
+  '.eslintrc.cjs',
+  '.eslintrc.yaml',
+  '.eslintrc.yml',
+  '.eslintrc.json',
+  '.eslintrc',
+];
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts:30]: Was there nothing we could borrow from eslint to read configs?

File: packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  meta: TaskMetaData = {
+    taskName: 'eslint-summary',
+    friendlyTaskName: 'Eslint Summary',
+    taskClassification: {
+      category: Category.Insights,
+      priority: Priority.Medium,
+    },
+  };
+
+  private eslintParser: Parser<ESLintReport>;
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts:42]: We should try to underscore private stuff, or use private fields.

File: packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    super(pluginName, context);
+
+    let createEslintParser = this.context.parsers.get('eslint')!;
+
+    let eslintConfig = readEslintConfig(
+      this.context.paths,
+      this.context.cliFlags.cwd,
+      this.context.pkg
+    );
+    this.eslintParser = createEslintParser(eslintConfig as ESLintOptions);
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts:55]: Why not give the var above a type, then you won't have to cast it.

File: packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    this.jsPaths = this.context.paths.filterByGlob('**/*.js');
+  }
+
+  async run(): Promise<TaskResult> {
+    let result: EslintSummaryTaskResult = new EslintSummaryTaskResult(this.meta);
+
+    let esLintReport = await this.runEsLint();
+    this.debug('ESLint Report', esLintReport);
+    result.esLintReport = esLintReport;
+    return result;
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts:66]: Nit: breathing room :)

File: packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  if (eslintConfigFile) {
+    return eslintConfigFile;
+  } else if (pkg.eslintConfig !== null) {
+    return pkg.eslintConfig as string;
+  } else {
+    throw new Error(
+      'No eslint config found in root (in the form of .eslintrc.* or as an eslintConfig field in package.json)'
+    );
+  }
+}
```
Comment (scalvert) [inline packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts:95]: Is there nothing we can use from eslint itself to read configs?

File: packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts (hunk 3 above)
Comment (carakessler) [reply packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:21]: yea its definitely something that has been top of mind to me as i have written a bunch of tasks this week.. i honestly have seen too many patterns emerge from these various data transformations, but they are definitely the brunt of the work when it comes to writing a task rn

File: packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts (hunk 6 above)
Comment (carakessler) [reply packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts:30]: the only thing i could find exported was `getConfigForFile` which obviously doesnt work for our use case

File: packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts (hunk 10 above)
Comment (carakessler) [reply packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts:95]: the only thing i could find exported was getConfigForFile which obviously doesnt work for our use case

File: packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts (hunk 8 above)
Comment (carakessler) [reply packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts:55]: we can give the var above a type, and i can cast in the `readEslintConfig` function and have that return `ESLintOptions`, but reading the file is going to return a string, and at some point i need to cast that to `ESLintOptions`, no?

File: packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts (hunk 2 above)
Comment (scalvert) [reply packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:15]: Yep. This is why I think it's useful for us to start thinking about schemas for the JSON we're producing in our tasks. Right now we're not adhering to any sort of schema, but rather just outputting what we think we want. We should probably do a future pass over all the output and figure out if we can provide some consistency.

File: packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts (hunk 8 above)
Comment (scalvert) [reply packages/checkup-plugin-javascript/src/tasks/eslint-summary-task.ts:55]: Ya I think if you have a function, like above, that reads an eslint config and returns a structured type, that function should just return the type itself. That way consumers won't have to guess at the structure - the type will dictate that.

File: packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts (hunk 2 above)
Comment (carakessler) [reply packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:15]: Adding to meeting agenda

File: packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts (hunk 4 above)
Comment (carakessler) [reply packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:62]: i am using the types provided by eslint, you can click into this and see that the severity is `type Severity = 0 | 1 | 2;`, but it doesnt say anywhere in the eslint types that 1 corresponds with warning :( read that in the docs

Comment (scalvert) [reply packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:62]: I think we define that type, don't we? If so, we can define 0, 1, 2 to mean the corresponding `off`, `warning`, `error` in eslint, no?

Comment (carakessler) [reply packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:62]: no, we are getting that type from ESlint

Comment (scalvert) [reply packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:62]: I was thinking of this https://github.com/checkupjs/checkup/blob/master/packages/core/src/types/ember-template-lint.ts#L1

Comment (scalvert) [reply packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:62]: Also, this file isn't importing/using any severity type. What type are you talking about? My proposal is to define `Severity`, or use the type above, and use that for comparison.

```
if (msg.severity === Severity.warning) {
  // ...
}
```

Comment (scalvert) [reply packages/checkup-plugin-javascript/src/results/eslint-summary-task-result.ts:62]: Let's just leave this for now.

**Pull request merged**